lib/spack/spack/config.py

```diff
@@ -779,31 +786,105 @@ def _config():
     """
     cfg = Configuration()
 
+    cfg_paths = cfg_paths or default_config_paths
+    install_tree = install_tree or spack.store.install_root
+
     # first do the builtin, hardcoded defaults
     defaults = InternalConfigScope('_builtin', config_defaults)
     cfg.push_scope(defaults)
 
     # add each scope and its platform-specific directory
-    for name, path in configuration_paths:
+    for name, path in cfg_paths.universal.items():
         cfg.push_scope(ConfigScope(name, path))
 
         # Each scope can have per-platfom overrides in subdirectories
         _add_platform_scope(cfg, ConfigScope, name, path)
 
+    shared_install_trees = cfg.get('config:shared_install_trees')
+    shared_tree_scope = False
+    if shared_install_trees:
+        if install_tree in shared_install_trees:
+            shared_tree_scope = True
+            scope_name = 'install_tree:' + install_tree
+            shared_install_root = shared_install_trees[install_tree]['root']
+            shared_install_root = spack.util.path.canonicalize_path(
+                shared_install_root)
+            shared_install_config_path = os.path.join(
+                str(shared_install_root), 'config')
+            cfg.push_scope(
+                ConfigScope(scope_name, shared_install_config_path))
+        else:
+            # User configuration is only used when we are not using a shared
+            # install tree
+            cfg.push_scope(ConfigScope(*cfg_paths.user))
+    else:
+        # User configuration is only used when we are not using a shared
+        # install tree
+        cfg.push_scope(ConfigScope(*cfg_paths.user))
+
     # add command-line scopes
     _add_command_line_scopes(cfg, command_line_scopes)
 
     # we make a special scope for spack commands so that they can
     # override configuration options.
     cfg.push_scope(InternalConfigScope('command_line'))
 
+    # TODO: this allows command-line scopes to specify new install trees.
+    # The added install_tree_scope needs to be lower precedence than all
+    # command line scopes.
+    if not shared_tree_scope:
+        # The user configuration will store the install trees that are used
+        # by this spack instance
+        install_trees = cfg.get('config:install_trees')
+        if not install_trees:
+            if shared_install_trees:
+                # If this is a shared spack instance, then the default for
+                # users is to put new installations in the home directory
+                install_trees = {'user': '~/.spack/installs'}
+            else:
+                install_trees = {'spack-root': '$spack/opt/spack'}
+            cfg.set('config:install_trees', install_trees, 'user')
+
+        default_install_tree = (not install_tree)
+        if not install_tree:
+            if shared_install_trees:
+                install_tree = 'user'
+            else:
+                install_tree = 'spack-root'
```

Comment:
Todd and I were talking: if the user replaces all the predefined install trees with their own (using different names) and then runs Spack without setting an `--install-tree`, this would result in a key error. At this point we either want to choose one of the user's `install_trees` by default, or raise an error (i.e. tell them to specify one).

Reply:
I think that in a scenario like this, raising an error is the right approach. I'll make that change.

Reply:
I have added an error for this case that informs the user that an install-tree must be specified when no default remains.